opensearch/src/main/java/org/opensearch/sql/opensearch/request/OpenSearchRequestBuilder.java

```diff
@@ -307,13 +328,41 @@ private void buildInnerHit(List<String> paths, NestedQueryBuilder query) {
     ));
   }
 
+  /**
+   * We need to group nested queries with same path for adding new fields with same path of
+   * inner hits. If we try to add additional inner hits with same path we get an OS error.
+   * @param nestedQueries Current list of nested queries in query.
+   * @param path path comparing with current nested queries.
+   * @return Query with same path or new empty nested query.
+   */
+  private NestedQueryBuilder findNestedQueryWithSamePath(
+      List<NestedQueryBuilder> nestedQueries, String path
+  ) {
+    return nestedQueries.stream()
+        .filter(query -> isSamePath(path, query))
+        .findAny()
+        .orElseGet(createEmptyNestedQuery(path));
+  }
+
+  /**
+   * Check is nested query of the same path value.
+   * @param path Value of path to compare with nested query.
+   * @param query nested query builder to compare with path.
+   * @return true if nested query has same path.
+   */
+  private boolean isSamePath(String path, NestedQueryBuilder query) {
+    return nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+  }
+
   /**
    * Create a nested query with match all filter to place inner hits.
    */
-  private NestedQueryBuilder createEmptyNestedQuery(String path) {
-    NestedQueryBuilder nestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
-    ((BoolQueryBuilder) query().filter().get(0)).must(nestedQuery);
-    return nestedQuery;
+  private Supplier<NestedQueryBuilder> createEmptyNestedQuery(String path) {
+    return () -> {
+      NestedQueryBuilder nestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
+      ((BoolQueryBuilder) query().filter().get(0)).must(nestedQuery);
+      return nestedQuery;
```

Comment:
What does this lambda function do? How is it different from a without the lambda function around it such as:
```
NestedQueryBuilder nestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
      ((BoolQueryBuilder) query().filter().get(0)).must(nestedQuery);
      return nestedQuery;
```

Reply:
Good question. The `Supplier` is an interface to fulfill the `orElseGet()` call in `findNestedQueryWithSamePath()`. We don't want to use `orElse()` and not use the `Supplier` cause that would have the `createEmptyNestedQuery()` code always execute. Instead we use `orElseGet()` which takes the `Supplier` interface and avoids unnecessary code execution with a non-empty optional from `findAny()`.

You can read more about this here:
https://www.baeldung.com/java-optional-or-else-vs-or-else-get